source/common/stats/stats_impl.cc

```diff
@@ -150,10 +150,20 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
 }
 
 RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  // This must be zero-initialized
   RawStatData* data = static_cast<RawStatData*>(::calloc(RawStatData::size(), 1));
   data->initialize(name);
-  return data;
+
+  std::unique_lock<std::mutex> lock(mutex_);
+  auto ret = stats_.insert(data);
+  lock.unlock();
```

Comment:
There's a subtle problem here.  The iterator you were returned can be invalidated if another element is inserted into the `set`.  You need to grab the raw pointer from the iterator while locked, and never use the iterator again after unlocking.

Reply:
Oh very good point Matt.  Just leave it locked till the end of the function then.  I can't think of why the hash implementation would need to invalidate the iterator but if the standard doesn't say it is safe then there is no point risking it.

Reply:
Oh very good point Matt.  Just leave it locked till the end of the function then.  I can't think of why the hash implementation would need to invalidate the iterator but if the standard doesn't say it is safe then there is no point risking it.

Reply:
I think the lock tightening is a good idea and will probably save us some cycles.  We just need to extract the raw pointer from the iterator while locked and only use the pointer temporary after we unlock.

As for how this interacts with a custom hash, my basic understanding is that as the set grows, it will at some point decide to rehash (using the same hash function) the entire set onto a larger set of buckets.  This is the only process that causes iterators to be invalidated for `std::unordered_set`.